File: skill_sdk/tests/fixtures/go-project/shop/payment.go
package shop

import "fmt"

// PaymentResult represents the outcome of a payment attempt.
type PaymentResult struct {
	Success       bool
	TransactionID string
	Message       string
}

// PaymentGateway defines the payment processing contract.
// goToImplementation on PaymentGateway should resolve to StripeGateway, PayPalGateway, MockGateway.
type PaymentGateway interface {
	Charge(amount float64, currency string) (*PaymentResult, error)
	Refund(transactionID string) (*PaymentResult, error)
}

// StripeGateway implements PaymentGateway via Stripe.
type StripeGateway struct {
	apiKey string
}

// NewStripeGateway creates a StripeGateway.
func NewStripeGateway(apiKey string) *StripeGateway {
	return &StripeGateway{apiKey: apiKey}
}

// Charge processes a payment through Stripe.
func (g *StripeGateway) Charge(amount float64, currency string) (*PaymentResult, error) {
	return &PaymentResult{
		Success:       true,
		TransactionID: fmt.Sprintf("stripe_ch_%d", hashFloat(amount)),
		Message:       fmt.Sprintf("Charged %.2f %s", amount, currency),
	}, nil
}

// Refund processes a refund through Stripe.
func (g *StripeGateway) Refund(transactionID string) (*PaymentResult, error) {
	return &PaymentResult{
		Success:       true,
		TransactionID: fmt.Sprintf("stripe_ref_%s", transactionID),
		Message:       fmt.Sprintf("Refunded transaction %s", transactionID),
	}, nil
}

// PayPalGateway implements PaymentGateway via PayPal.
type PayPalGateway struct {
	clientID     string
	clientSecret string
}

// NewPayPalGateway creates a PayPalGateway.
func NewPayPalGateway(clientID, clientSecret string) *PayPalGateway {
	return &PayPalGateway{clientID: clientID, clientSecret: clientSecret}
}

// Charge processes a payment through PayPal.
func (g *PayPalGateway) Charge(amount float64, currency string) (*PaymentResult, error) {
	return &PaymentResult{
		Success:       true,
		TransactionID: fmt.Sprintf("pp_ch_%d", hashFloat(amount)),
		Message:       fmt.Sprintf("PayPal charge %.2f %s", amount, currency),
	}, nil
}

// Refund processes a refund through PayPal.
func (g *PayPalGateway) Refund(transactionID string) (*PaymentResult, error) {
	return &PaymentResult{
		Success:       true,
		TransactionID: fmt.Sprintf("pp_ref_%s", transactionID),
		Message:       fmt.Sprintf("PayPal refund %s", transactionID),
	}, nil
}

// MockGateway implements PaymentGateway for testing.
type MockGateway struct {
	callCount int
}

// NewMockGateway creates a MockGateway.
func NewMockGateway() *MockGateway {
	return &MockGateway{}
}

// Charge simulates a payment.
func (g *MockGateway) Charge(amount float64, currency string) (*PaymentResult, error) {
	g.callCount++
	return &PaymentResult{
		Success:       true,
		TransactionID: fmt.Sprintf("mock_tx_%d", g.callCount),
		Message:       fmt.Sprintf("Mock charge %.2f %s", amount, currency),
	}, nil
}

// Refund simulates a refund.
func (g *MockGateway) Refund(transactionID string) (*PaymentResult, error) {
	return &PaymentResult{
		Success:       true,
		TransactionID: fmt.Sprintf("mock_ref_%s", transactionID),
		Message:       fmt.Sprintf("Mock refund %s", transactionID),
	}, nil
}

func hashFloat(f float64) int {
	return int(f * 100)
}

File: skill_sdk/tests/fixtures/py-project/shop/payment.py
"""Payment gateway abstraction — ABC + strategies for implementations."""

from __future__ import annotations

from abc import ABC, abstractmethod
from dataclasses import dataclass
from typing import List


@dataclass
class PaymentResult:
    """Outcome of a payment attempt."""
    success: bool
    transaction_id: str
    message: str


class PaymentGateway(ABC):
    """Abstract payment processor. goToImplementation resolves concrete classes."""

    @abstractmethod
    def charge(self, amount: float, currency: str) -> PaymentResult:
        ...

    @abstractmethod
    def refund(self, transaction_id: str) -> PaymentResult:
        ...


class StripeGateway(PaymentGateway):
    """Stripe-backed payment gateway implementation."""

    def __init__(self, api_key: str) -> None:
        self._api_key = api_key

    def charge(self, amount: float, currency: str) -> PaymentResult:
        return PaymentResult(
            success=True,
            transaction_id=f"stripe_ch_{hash(str(amount))}",
            message=f"Charged {amount} {currency}",
        )

    def refund(self, transaction_id: str) -> PaymentResult:
        return PaymentResult(
            success=True,
            transaction_id=f"stripe_ref_{transaction_id}",
            message=f"Refunded transaction {transaction_id}",
        )


class PayPalGateway(PaymentGateway):
    """PayPal-backed payment gateway implementation."""

    def __init__(self, client_id: str, client_secret: str) -> None:
        self._client_id = client_id
        self._client_secret = client_secret

    def charge(self, amount: float, currency: str) -> PaymentResult:
        return PaymentResult(
            success=True,
            transaction_id=f"pp_ch_{hash(str(amount))}",
            message=f"PayPal charge {amount} {currency}",
        )

    def refund(self, transaction_id: str) -> PaymentResult:
        return PaymentResult(
            success=True,
            transaction_id=f"pp_ref_{transaction_id}",
            message=f"PayPal refund {transaction_id}",
        )


class MockGateway(PaymentGateway):
    """Always-succeeds mock — useful for testing."""

    def __init__(self) -> None:
        self._call_count = 0

    def charge(self, amount: float, currency: str) -> PaymentResult:
        self._call_count += 1
        return PaymentResult(
            success=True,
            transaction_id=f"mock_tx_{self._call_count}",
            message=f"Mock charge {amount} {currency}",
        )

    def refund(self, transaction_id: str) -> PaymentResult:
        return PaymentResult(
            success=True,
            transaction_id=f"mock_ref_{transaction_id}",
            message=f"Mock refund {transaction_id}",
        )
